orangecrab/src/Ila.hs

```diff
@@ -73,26 +78,89 @@ triggerController ::
   -- | Circuit outputting the content of the buffer whenever the predicate has been triggered
   -- The input signal is to clear the trigger
   Circuit
-    (CSignal dom Bool)
+    (CSignal dom (Maybe (Index size)), CSignal dom Bool)
     (PacketStream dom (BitSize a `DivRU` 8) IlaDataHeader)
-triggerController predicate i core = Circuit exposeIn
+triggerController config predicate core = Circuit exposeIn
  where
-  exposeIn (triggerRst, backpressure) = out
+  exposeIn ((newTrigPoint, triggerRst), backpressure) = out
    where
-    oldTriggered :: Signal dom Bool
-    oldTriggered = register False triggered
+    triggerPoint = register config.triggerPoint $ liftA2 DM.fromMaybe triggerPoint newTrigPoint
+
+    postTriggerSampled :: Signal dom (Index size)
+    postTriggerSampled = register 0 $ mux triggered (satAdd SatBound 1 <$> postTriggerSampled) 0
+
     triggered :: Signal dom Bool
-    triggered = mux triggerRst (pure False) oldTriggered .||. (predicate <$> i)
+    triggered =
+      register False $ mux triggerRst (pure False) triggered .||. (predicate <$> config.tracing)
 
     shouldSample :: Signal dom Bool
-    shouldSample = not <$> triggered
+    shouldSample = not <$> triggered .||. postTriggerSampled .<. triggerPoint
 
-    injectId oldMeta = (0, oldMeta)
+    injectId oldMeta = (config.hash, oldMeta)
 
-    buffer = core i shouldSample triggerRst
-    Circuit packet = dataPacket (Proxy :: Proxy a) <| mapMeta injectId <| ringBufferReaderPS buffer
+    buffer = core config.tracing shouldSample triggerRst
+    Circuit packet = dataPacket <| mapMeta injectId <| ringBufferReaderPS buffer
 
-    out = (pure (), snd $ packet (triggered, backpressure))
+    out = ((pure (), pure ()), snd $ packet (triggered, backpressure))
+
+-- | A signal fanout, simply duplicates a signal n times
+fanoutCSig ::
+  forall dom n a.
+  (KnownNat n) =>
+  -- | How many times to 'duplicate' the signal
+  SNat n ->
+  -- | The circuit, where the input will be duplicated n times and returned as the output
+  Circuit
+    (CSignal dom a)
+    (Vec n (CSignal dom a))
+fanoutCSig _ = Circuit go
+ where
+  go (fwd, _) = (pure (), repeat fwd)
```

Comment:
You should not need this anymore know that we know how to use the `Fwd` pattern match and Circuit construction